aten/src/ATen/native/LinearAlgebra.cpp

```diff
@@ -1852,6 +1811,89 @@ static Tensor& linalg_norm_out_impl(Tensor& result, const Tensor& self, optional
   return result;
 }
 
+static Tensor& linalg_vector_norm_impl(Tensor& result, const Tensor& self, optional<Scalar> opt_ord, optional<IntArrayRef> opt_dim, bool keepdim, optional<ScalarType> opt_dtype) {
+  auto ord = opt_ord.value_or(2).toDouble();
+
+  TORCH_CHECK(self.device().type() == DeviceType::CPU || self.device().type() == DeviceType::CUDA,
+              "norm only supports CPU AND CUDA device type, got: ", self.device().type());
+  TORCH_CHECK(self.layout() == Layout::Strided,
+              "norm only supports strided layout, got: ", self.layout());
+
+  ScalarType in_dtype = opt_dtype.has_value() ? opt_dtype.value() : self.scalar_type();
+  TORCH_CHECK(
+      at::isFloatingType(in_dtype) || at::isComplexType(in_dtype),
+      "Can only calculate the norm of floating point and complex dtypes. Got ",
+      toString(in_dtype),
+      " instead.");
+
+  IntArrayRef dim = opt_dim.value_or(IntArrayRef{});
+
+  // For NumPy compatibility, we can only perform order +/-infinity reduction
+  // (max/min) on a tensor with zero elements if the dimensions to reduce are
+  // nonzero. Otherwise, throw an error.
+  if (self.numel() == 0 && std::abs(ord) == INFINITY) {
+    bool has_identity = true;
+
+    if (dim.size() == 0) {
+      has_identity = false;
+    } else {
+      for (int64_t dim_num : dim) {
+        if (self.size(dim_num) == 0) {
+          has_identity = false;
+          break;
+        }
+      }
+    }
+    if (!has_identity) {
+      TORCH_CHECK(ord != INFINITY, (
+        "cannot perform reduction function max on tensor with "
+        "no elements because the operation does not have an identity"));
+      TORCH_CHECK(ord != -INFINITY, (
+        "cannot perform reduction function min on tensor with "
+        "no elements because the operation does not have an identity"));
+    }
+  }
+
+  ScalarType out_dtype = opt_dtype.has_value() ? opt_dtype.value() : toValueType(self.scalar_type());
+  Tensor result_;
+  Tensor self_;
+
+  if (self.device().type() == c10::kCPU && std::abs(ord) == INFINITY) {
+    // TODO: For some reason, for complex CPU, at::abs() gives a different
```

Comment:
I now understand why this happens.

First of all, it's worth mentioning, for clarity, that in order to find out which element from `self` was returned by the `vector_norm` call as the min/max value, the backward method performs `self.abs() == norm_result`. Then, it simply sets the corresponding element(s) in the gradient to 1, and all else to 0.

The root issue that causes the backward function to fail is `self.abs() == norm_result`. `linalg_vector_norm`'s dispatch function `norm_stub` with infinite order is using a per-element `std::abs` call before applying reductions, instead of using `at::abs`. Thus it becomes possible for none of the elements from `self.abs()` to match exactly with `norm_result`. The reason they could mismatch is because `at::abs` is free to use a vectorized abs function, which can produce slightly different results than `std::abs`.

Below is an example of a situation where vectorized and nonvectorized `torch.abs` give different results with a complex input (at least in my environment):

```
import torch
b = torch.tensor([(7.652771299449274e-06-1.7120581036111606e-06j), (1.9727849811062448e-05+1.919867619816995e-05j), (1.6650342521593928e-05-1.0042102916947193e-05j), (-4.901525595762313e-06-6.710164915302073e-06j)], dtype=torch.complex128)
print(b.abs().tolist())
```

Run with `ATEN_CPU_CAPABILITY=default` to disable vectorization, it prints:
```
[7.841941820226363e-06, 2.7527753739272188e-05, 1.9444221174450732e-05, 8.309709161966138e-06]
```

Run with `ATEN_CPU_CAPABILITY=avx2` to enable vectorization, it prints:
```
[7.841941820226363e-06, 2.752775373927219e-05, 1.944422117445073e-05, 8.309709161966138e-06]
```

Notice that the second and third values in both outputs are slightly different. The backward `at::abs()` call is free to be vectorized, so it can disagree with the `std::abs` per-element call within `norm_stub`.

The workaround that I added here simply calls `at::abs` if an infinite order was given to the CPU implementation, so it avoids the complex per-element `std::abs` call in `norm_stub`. The forward `at::abs` call should give the exact same result as the backward `at::abs` call. But that is not ideal since dispatching to `at::abs()` has some extra overhead that I would rather avoid if possible.

Furthermore, the strategy that `linalg_vector_norm_backward` uses to determine which element was chosen from the input is also not ideal. It takes extra time to recalculate information (`self.abs()`) that could potentially be captured by the forward function. But as far as understand, in order to do that we would need to add an extra return value (in a `std::tuple`) to `linalg_vector_norm`. Or is there some other way to do it? I'm going to look into how `torch.amax/amin` do this--perhaps they use a good strategy that I can mimic.

Reply:
@ngimel or @mruberry, do have some insight to what I should do about this problem? Apologies if my description is unclear, let me know if so.

Reply:
How's current `norm` dealing with this problem?

Reply:
It doesn't. `linalg_vector_norm_backward` and `norm_backward` use the same technique for infinite orders. If I add an OpInfo for `torch.norm`, it fails in the same way that `torch.linalg.vector_norm` fails. `torch.linalg.norm`, on the other hand, passes in this case in the current main branch, because it's directly calling `at::abs` in the forward: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/LinearAlgebra.cpp#L1694

Reply:
A possible solution could be to vectorize the CPU implementation of `vector_norm`. As long as vectorization for the absolute value calculation is applied in the same way that `at::abs()` does it, then the backward and forward calls would match. This should also improve performance a fair bit.

Reply:
cc @albanD to also weigh in on this issue, and I've added it to our sync discussion for tomorrow as well

Reply:
@albanD @ngimel and I had an opportunity to discuss and we'd like to propose keeping the working of unfused abs in forward as long as it's limited to cpu complex and inf order. Would you file a follow-up issue, however, @kurtamohler, describing this challenge? We would like to consider more significant changes to improve the performance here in the future.

Reply:
Filed #52648